testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/ComponentsTest.java

```diff
@@ -172,6 +172,25 @@ public void testCreateWithGivenId() {
         assertEquals("fixed-id", id);
     }
 
+    @Test
+    public void failCreateWithLongName() {
+        StringBuilder name = new StringBuilder();
+
+        while (name.length() < 300) {
+            name.append("invalid");
+        }
+
+        ComponentRepresentation rep = createComponentRepresentation(name.toString());
+
+        rep.getConfig().addFirst("required", "foo");
+
+        ComponentsResource components = realm.components();
+
+        try (Response response = components.add(rep)) {
+            assertEquals(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), response.getStatus());
+        }
```

Comment:
The problem is that the component is not persisted even when 201 is returned due to constraint validation errors when committing the transaction. And this response is inconsistent with the Wildfly distribution because of the new request lifecycle.

In fact, we have also a difference between the old and the map store here. Using a database should always fail with 500 but with the map store, it always succeeds because there is no constraint/validation for the component name.

Looks like you are saying that the map store behavior is expected and we should just consider both behaviors to fix the failing test?

Reply:
Let's take a step back and clarify two things:

What is the expected behaviour? Is it the following or something different?
- When component is created, return 201
- When component is not created, return 500

> Using a database should always fail with 500

Why?

Reply:
The behavior I expect is a proper input validation and return 400 instead. But considering how it works today, that is the behavior. I'm not changing this here.

The database should always fail because of the constraints we have in tables. See the issue for the stack trace.

Is the map store enforcing the same constraints ?

Reply:
Thanks, now it makes sense.

Map store is an abstraction so actual constraints are not imposed by map storage itself, but according to the map storage implementation being effective during that run.

In the case of this GHA undertow-map, `ConcurrentHashMap` implementation is used which has no constraints at all, and that is the reason why the component is correctly created and 200 is returned. If a JPA / map would be used, the constraints on component names would probably not be there since the component definitions would be part of a realm JSON, so it would also pass. There might be some storages that would limit component names in the future, these would fail to save such a component.

Hence I think the test checking the (non)existence of a component per 201 / 500 status being the best what can be done at this moment. WDYT?

Reply:
Thanks. And that is why I asked you before if we could just ignore that particular test method for the map store.

Or even add a condition to cover both behaviors (expected and map store).

Reply:
The issue as far as I understand it is to fix the difference in behaviour between quarkus and wildfly response operating on the same data store. Wildfly with map store / CHM would also return 201 and create the component correctly. It thus cannot be conditioned on map store or other store, but can be on the response code (201 or 4xx / 5xx) and checking the existence of the component.

@DGuhr Store implementation can be constrained by the technology which is the case of the legacy database. Ideal behaviour would be to allow strings of any length. 255 characters are often picked not only due to indexing but also for RDBMS storage savings since then the string length fits one byte. I thus don't follow the argument about "against `ConcurrentHashMap`". The argument would go that CHM behaves better than RDBMS that is limited in the component name length.

Reply:
@DGuhr @hmlnarik I'm changing the test to cover both storages. If the new behavior is expected considering the new store design, I'm OK.

Reply:
Ok - @hmlnarik @pedroigor Let me try to explain:

I thought about expectations and surprises here when changing the dev db to map.

Context: e.g. in the testsuite, when using the IDELauncherand with devmode itself invoking it from the cli we run `start-dev` under the covers. Changing the default Database now from h2 (with the constraints) to map (without constraints) would cause some surprises i guess, when you accidentally violate a constraint for, say, postgres, but you won't notice until it's run with an actual postgres, or a db that has these constraints. So imo it's at least discussable to have a non-constrained datastore for this, as this could cause surprises very late in the flow of developing.

@stianst Do we have a discussion around moving dev-file from h2 to...? I think this here is not exxactly the place to discuss this topic, nevertheless should be discussed ;)

Reply:
+1 for a separate discussion.

Short note: that this is by no means different from current state where [some database surprised us](https://issues.redhat.com/browse/KEYCLOAK-4122) late in the cycle due to their specifics.

Reply:
The expected status codes is whatever HTTP specs say. We often result in 500's today due to invalid request, which is just not correct. Reason for this is often down to missing validation, but also in terms of when the error comes from the DB we're not able to map the SQL exceptions properly to a proper status code. Part of the solution to that is better validation, which I believe has to be a thing done by the new store, as there is some validation that can be done upfront, but then there's also store-specific limitations/validations I presume.

The point about map store not being a good fit as a replacement for dev-file/mem as it doesn't verify if a particular database can store it is not a valid argument at all. Testing with H2 in dev, does not give you any guarantees that it will work with PostgreSQL in production. Besides, the plan is also to offer static store as an option in production as well.

I'd suggest we separate this into a discussion around validation, and expected status codes to be returned under different situations. Now, as no-one is going to start working on that right now I would suggest we table this discussion a little bit until we have identified what team should work on validation (hint: I will push hard on extending/unifying our validation after the new store is ready).